File: Niveau 3/Niveau 3.7/cavernes.py
#-------------------------------------------------------------------------------
# Name:        Carte des cavernes
# Purpose:     France-IOI niveau 3.7
#
# Author:      Sébastien
#
# Created:     26/09/2022
#-------------------------------------------------------------------------------

def valide(s):
    compteur = 0
    for c in s:
        if c == '(':
            compteur += 1
        elif c == ')':
            compteur -= 1
        if compteur < 0:
            return False
    return compteur == 0

def main():
    n = int(input())
    s = input()
    if valide(s):
        print(1)
    else:
        print(0)

if __name__ == '__main__':
    main()

"""
18
( ( () (()) ) () )
----
2
)(
----
9
( () (( )
----
0

"""

File: Niveau 3/Niveau 3.7/cavernes.hs
--------------------------------------------------------------------------------
-- Name:        Carte des cavernes
-- Purpose:     France-IOI niveau 3.7
--
-- Author:      Sébastien
--
-- Created:     26/09/2022
--------------------------------------------------------------------------------

valide :: String -> Int -> Bool
valide "" c = (c == 0) -- Fin de chaîne
valide ('(':xs) c = valide xs (c+1)
valide (')':xs) c = if c > 0 then valide xs (c-1) else False
valide (_:xs) c = valide xs c -- Ignore les caractères autres que ( ou )

main :: IO()
main = do  
    _ <- getLine
    s <- getLine
    putStrLn $ if valide s 0 then "1" else "0"
    
